In [138]:
import pandas as pd
import numpy as np
import datetime as dt
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns
import operator
from functools import reduce

In [139]:
#cd C:\Users\bdhut\Desktop\Python Scripts

In [140]:
#pwd

The above code doesn't seem to change where the file is saved. 

Stack Overflow is designed to answer questions based on programming. Users are encouraged to search for questions before asking them to avoid repeats. 

https://data.stackexchange.com/datascience/query/edit/1410799

This link provides SQL-like querying of posts, flags, tags, and other data available through StackExchange website. Specifically to the Data Science section. 

In [141]:
df = pd.read_csv('Stackexquery.csv')

In [142]:
df.head()

Id  PostTypeId         CreationDate  Score  ViewCount  \
0  65650           2  2019-12-30 23:35:48      0        NaN   
1  65649           1  2019-12-30 22:37:09      3       91.0   
2  65648           1  2019-12-30 22:10:01      2      118.0   
3  65647           2  2019-12-30 22:06:00      0        NaN   
4  65646           2  2019-12-30 20:43:55      0        NaN   

                                                Tags  AnswerCount  \
0                                                NaN          NaN   
1  <python><neural-network><keras><loss-function>...          0.0   
2  <python><deep-learning><keras><reinforcement-l...          2.0   
3                                                NaN          NaN   
4                                                NaN          NaN   

   FavoriteCount  
0            NaN  
1            NaN  
2            2.0  
3            NaN  
4            NaN

In [143]:
df[['Id','PostTypeId','CreationDate','Score','ViewCount','Tags','AnswerCount','FavoriteCount']].isnull().sum()

Id                   0
PostTypeId           0
CreationDate         0
Score                0
ViewCount         7890
Tags              7890
AnswerCount       7890
FavoriteCount    13015
dtype: int64

ViewCount, Tags, and AnswerCount all have 7890 null values.

In [144]:
df.dtypes

Id                 int64
PostTypeId         int64
CreationDate      object
Score              int64
ViewCount        float64
Tags              object
AnswerCount      float64
FavoriteCount    float64
dtype: object

In [145]:
df.head()

Id  PostTypeId         CreationDate  Score  ViewCount  \
0  65650           2  2019-12-30 23:35:48      0        NaN   
1  65649           1  2019-12-30 22:37:09      3       91.0   
2  65648           1  2019-12-30 22:10:01      2      118.0   
3  65647           2  2019-12-30 22:06:00      0        NaN   
4  65646           2  2019-12-30 20:43:55      0        NaN   

                                                Tags  AnswerCount  \
0                                                NaN          NaN   
1  <python><neural-network><keras><loss-function>...          0.0   
2  <python><deep-learning><keras><reinforcement-l...          2.0   
3                                                NaN          NaN   
4                                                NaN          NaN   

   FavoriteCount  
0            NaN  
1            NaN  
2            2.0  
3            NaN  
4            NaN

In [146]:
df['CreationDate'] = pd.to_datetime(df['CreationDate'])

In [147]:
df.dtypes

Id                        int64
PostTypeId                int64
CreationDate     datetime64[ns]
Score                     int64
ViewCount               float64
Tags                     object
AnswerCount             float64
FavoriteCount           float64
dtype: object

In [148]:
df['Tags']=df['Tags'].str.replace('><',',')

In [149]:
df['Tags']=df['Tags'].str.replace('>','')

In [150]:
df['Tags']=df['Tags'].str.replace('<','')

In [151]:
df['Tags']=df['Tags'].str.split(',')

I think I could've combined a few of lthose lines ex .str.replace(...).str.replace(...)

In [152]:
df.head()

Id  PostTypeId        CreationDate  Score  ViewCount  \
0  65650           2 2019-12-30 23:35:48      0        NaN   
1  65649           1 2019-12-30 22:37:09      3       91.0   
2  65648           1 2019-12-30 22:10:01      2      118.0   
3  65647           2 2019-12-30 22:06:00      0        NaN   
4  65646           2 2019-12-30 20:43:55      0        NaN   

                                                Tags  AnswerCount  \
0                                                NaN          NaN   
1  [python, neural-network, keras, loss-function,...          0.0   
2  [python, deep-learning, keras, reinforcement-l...          2.0   
3                                                NaN          NaN   
4                                                NaN          NaN   

   FavoriteCount  
0            NaN  
1            NaN  
2            2.0  
3            NaN  
4            NaN

In [153]:
df.dtypes

Id                        int64
PostTypeId                int64
CreationDate     datetime64[ns]
Score                     int64
ViewCount               float64
Tags                     object
AnswerCount             float64
FavoriteCount           float64
dtype: object

In [154]:
df = df.fillna(0)

In [155]:
test = df['Tags']

In [156]:
test[8]

['machine-learning', 'regression', 'rnn', 'model-selection', 'k-nn']

In [157]:
tags_used_dict2 ={}
for tag in test[8]:
    if not tag in tags_used_dict2:
        tags_used_dict2[tag]=0
    tags_used_dict2[tag]+=1

In [158]:
tags_used_dict2

{'machine-learning': 1,
 'regression': 1,
 'rnn': 1,
 'model-selection': 1,
 'k-nn': 1}

In [159]:
# Creating a dictionary for all the tags
tags_used_dict = {}
for lst in test:
    if lst == 0:
        continue
    for tag in lst:
        if not tag in tags_used_dict:
            tags_used_dict[tag] = 0
        tags_used_dict[tag] += 1   
        
# Defining a function for creating a dataframe from a dictionary, 
# sorted in descending order
def create_sorted_df_from_dict(dictionary, new_df_column_name):
    
    # Creating a dataframe from the dictionary
    tags2 = pd.DataFrame.from_dict(dictionary, orient='index').reset_index() 

    # Renaming columns
    tags2.columns = ['Tag', new_df_column_name]

    # Sorting in descending order
    tags2 = tags2.sort_values(new_df_column_name, ascending=False).reset_index(drop=True)
    return tags2, tags2.head(10)

# Creating 2 dataframes: tags with the corresponding counts and the TOP10 used tags
tags_used, top10_used = create_sorted_df_from_dict(
                                                   dictionary=tags_used_dict,
                                                   new_df_column_name='Count'
                                                   )  

top10_used

Tag  Count
0  machine-learning   2133
1            python   1446
2     deep-learning    893
3    neural-network    819
4             keras    705
5    classification    578
6      scikit-learn    447
7               nlp    444
8        tensorflow    414
9               cnn    378